### PR TITLE
#303: Ensure consistency with Javaslang-Jackson serialization/deserialization

### PR DESCRIPTION
**Pull request by Diagoras**
Making this module able to deserialize Javaslang's Either.

### CODE REVIEW

**Comment by Diagoras** [comment]
Anything else needed to get this PR to a mergeable state? AFAICT, the Travis CI failure is because the Javaslang library requires Java 8, but we only use it as a test dependency. Is there an easy way around those failures?

**Comment by nbauernfeind** [comment]
It feels to me that this logic should sit in a module of its own. Is the issue that jackson finds the first thing that knows how to deserialize `Either` and that happens to be the jackson module not your custom javaslang one? Did you try such an approach?

As for the test failure (no need to fix yet, because I am not convinced this should be merged) -- the easiest way around it would be to create an object in the test that most closely represents the javaslang object. This gives you the best of both worlds (loose-coupling between the libraries, but retain the ability to execute tests against the other's type structure). It is also possible to stuff that all into a special module that will only be compiled/packaged/tested if a sufficiently new version of java is being used.

Is this change still something you would like to see integrated in some fashion, or have you already moved on (sorry for the late response...)?

**Comment by Diagoras** [comment]
@nbauernfeind [The related issue](https://github.com/FasterXML/jackson-module-scala/issues/303) provides the full context, but to summarize: the Javaslang Jackson module serializes an Either using "left" and "right", while this project prefers "l" and "r". That means JSON which represents a Scala Either cannot be deserialized into a Javaslang Either (or vice versa), which is annoying if you're trying to link together various Scala and Java microservices.

My preferred solution was to settle on a single standard, but both projects raised reasonable concerns about backwards compatibility. Everyone decided to go with the alternative of continuing to serialize Either differently, but to maintain mutual compatibility when deserializing by accepting both formats.

[The Jackson Javaslang module merged the code to allow compatible deserialization a while ago.](https://github.com/vavr-io/vavr-jackson/issues/98) Now we're just waiting on this PR before the two approaches are fully compatible.

Does that clear things up? Let me know if need any more details - and if there's some easier way to ensure that your Either can be read from a serialized Javaslang Either, let me know and I can implement that instead.

**Comment by nbauernfeind** [comment]
Very clear! Thanks. I'm going to modify the tests to be javaslang-less to avoid pulling in the dependency.

**Comment by Diagoras** [comment]
Makes sense. Let me know if you need a hand, or if you have any other questions.

**Comment by mbknor** [comment]
Hi..

sorry for not looking into this PR.

Rebase it on master and make it compile.... Then I'll have a look at it.

**Comment by cowtowncoder** [comment]
@Diagoras not sure if you are still interested in this, but sounds like this could be merged if conflicts could be resolved?

**Comment by Diagoras** [comment]
@cowtowncoder Yeah, I'm interested. Having a little trouble with my build process, though, since it's hard to find anywhere to get Java 7 for more recent versions of Ubuntu. Once I get that fixed, though, I should be able to test locally.

**Comment by cowtowncoder** [comment]
@Diagoras I think all builds should be using JDK 8 by now, FWTW. Also: @pjfanning may be able to help with build issues regarding `sbt`.

**Comment by pjfanning** [comment]
@Diagoras could you rebase this onto 2.10 branch? - master is a very experimental and broken branch targetted for the 3.0 release

**Comment by Diagoras** [comment]
@pjfanning  I think it should be good now. Builds and tests alright locally.

File: src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+  * Ensuring we can read serialized Vavr Either types into their equivalent Scala types.
+  */
+class VavrDeserializerTest extends DeserializerTest with EitherJsonTestSupport {
+
+  val module: DefaultScalaModule.type = DefaultScalaModule
+
+  "DefaultScalaModule" should "be able to deserialize right with string" in {
+    deserialize[Either[_, String]](s"""["right", "$str"]""")
```
Comment (pjfanning) [inline src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: why do we need to support either represented as an array when we already support the more logical `{"r":"$str"}`

Comment (Diagoras) [reply src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: @pjfanning Because that's how the Vavr Jackson module serializes their Either types, and this PR is about adding compatibility for handling that format. [This comment goes into some more detail about the purpose for this change](https://github.com/FasterXML/jackson-module-scala/pull/311#issuecomment-479933407), if you're interested.

Comment (pjfanning) [reply src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: that comment above does not explain why you want to support the json array format - I would be open to supporting `{"right":"$str"}` as well as `{"r":"$str"}` but I do not support the idea of supporting `["right", "$str"]` (or `["r", "$str"]`)

Comment (Diagoras) [reply src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: @pjfanning I think it does explain, but it sounds like I worded it badly. Let me try again.

[As you can see here](https://github.com/vavr-io/vavr-jackson/blob/master/src/main/java/io/vavr/jackson/datatype/serialize/EitherSerializer.java), the Vavr Jackson module serializes an Either as an array. That means to resolve issue #303 requires being able to handle deserializing from the array-style format.

The other alternative would be to close this issue and PR as "Won't Do", but I think that would be a bad idea for a number of reasons:

1. The maintainers of the Vavr Jackson module already [merged a similar PR](https://github.com/vavr-io/vavr-jackson/issues/98) that provided deserialization compatibility with the object-style format that this module produces. It was merged under the assumption that the Scala Jackson module would reciprocate, so refusing at this point might come as a surprise.
2. The maintainers of the Vavr Jackson module have expressed interest in using the object-style format instead, but think that handling the array-style format is needed for backwards compatibility. Might be worth pushing them to change, and then eventually reverting support for parsing array-style format a few versions from now.
3. We'd end up in a situation where people can write a Scala Either as JSON and have it read as a Vavr Either, but *could not* write a Vavr Either as JSON and have it read as a Scala Either. That'd be kinda confusing.
4. Last and least, everybody seemed onboard with this (Vavr and Jackson team) when this issue and PR was first opened back in 2017. Unless something changed in the preceding time, it'd have been nicer to object to and close this ticket before we did any work on it. @cowtowncoder, please keep me honest here - if you were okay with this ticket because you thought it was about something else, then that'd be on me instead.

Regardless, it's your call whether or not you'll add reciprocal support for handling Vavr's Either. No matter your decision, thanks for taking the time to review this and maintaining this kickass library.

**Comment by pjfanning** [comment]
https://github.com/FasterXML/jackson-module-scala/commit/543b10cfd69b02d4b44e9469b44387f2ef842049 provides some support

File: src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala (hunk 1 above)
Comment (cowtowncoder) [reply src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: I do not decide things on Scala module, and would defer to PJ, Nate and Morten -- all of who have commented at some point -- but I can see both sides. To me it seems reasonable to support alternative "legacy" format, with strict checking, but not produce it via serialization. Meaning that only very specific kind of array (2 elements, first as matching id) would be accepted, others rejected.
Personally I am more lenient in cases of interoperability; if it wasn't for that part, I would not like the idea of adding another representation. But due to existing usage, wish for backwards compatibility (even tho it is not for older Scala module), I think this would be ok.

@pjfanning does this make more sense?

Comment (pjfanning) [reply src/test/scala/com/fasterxml/jackson/module/scala/deser/VavrDeserializerTest.scala]: Merged - I confirmed that vavr uses that strange array representation for Either.
